File: app/src/main/java/jgappsandgames/smartreminderslite/adapter/TaskAdapterInterface.java
package jgappsandgames.smartreminderslite.adapter;

// Java
import java.util.ArrayList;

// Jetbrains
import org.jetbrains.annotations.Nullable;

// Android OS
import android.app.Activity;
import android.util.Log;

// Views
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

// App
import jgappsandgames.smartreminderslite.R;
import jgappsandgames.smartreminderslite.holder.TaskFolderHolder;

// Save
import jgappsandgames.smartreminderssave.tasks.Task;

/**
 * TaskAdapterInterface
 * Created by joshua on 10/22/17.
 *
 * The Main Adapter for any Task List
 * It Takes in an Array of either Task filepaths or Tasks and Displays them
 */
public abstract class TaskAdapterInterface extends BaseAdapter {
    // Log Constant
    private static final String LOG = "TaskAdapterInterface";

    // Task Type Constant
    private static final int TASK_COUNT = 3;

    // Data
    private final Activity activity;
    private final TaskFolderHolder.OnTaskChangedListener listener;
    private final ArrayList<Task> tasks;

    // Initializers
    protected TaskAdapterInterface(Activity activity, TaskFolderHolder.OnTaskChangedListener listener, ArrayList<Task> tasks) {
        Log.d(LOG, "Initializing.");
        this.activity = activity;
        this.listener = listener;
        this.tasks = tasks;

        Log.v(LOG, String.valueOf(this.tasks.size()));

        Log.v(LOG, "Finished Initializing.");
    }

    protected TaskAdapterInterface(Activity activity, TaskFolderHolder.OnTaskChangedListener listener, ArrayList<String> tasks, @SuppressWarnings({"SameParameterValue", "unused"}) @Nullable String unused) {
        Log.d(LOG, "Initializing.");
        this.activity = activity;
        this.listener = listener;

        this.tasks = new ArrayList<>();
        for (String t : tasks) this.tasks.add(new Task(t));

        Log.v(LOG, String.valueOf(this.tasks.size()));

        Log.v(LOG, "Finished Initializing.");
    }

    // List Methods
    @Override
    public int getCount() {
        return tasks.size();
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public int getViewTypeCount() {
        return TASK_COUNT;
    }

    // Item Methods
    @Override
    public Task getItem(int position) {
        return tasks.get(position);
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).getID();
    }

    @Override
    public int getItemViewType(int position) {
        return getItem(position).getType();
    }

    @Override
    public View getView(int position, View convert_view, ViewGroup parent) {
        Log.d(LOG, "Get View Called");
        TaskFolderHolder holder;

        // View Does Not Exist so Create it
        if (convert_view == null) {
            Log.v(LOG, "Create the View");
            if (getItem(position).getType() == Task.TYPE_FLDR) convert_view = LayoutInflater.from(activity).inflate(R.layout.list_folder, parent, false);
            else convert_view = LayoutInflater.from(activity).inflate(R.layout.list_task, parent, false);

            holder = new TaskFolderHolder(getItem(position), convert_view, activity, listener);
            convert_view.setTag(holder);
        }

        // View Does Exist so Just Get it
        else {
            Log.v(LOG, "Load the View");
            holder = (TaskFolderHolder) convert_view.getTag();
            holder.task = getItem(position);
        }

        // Update View
        Log.v(LOG, "Update the View");
        holder.setViews();

        // Return the View
        Log.v(LOG, "Return the View");
        return convert_view;
    }
}

File: app/src/main/java/jgappsandgames/smartreminderslite/adapter/TaskAdapterInterface.kt
package jgappsandgames.smartreminderslite.adapter

// Android OS
import android.app.Activity

// Views
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter

// Jetbrain
import org.jetbrains.annotations.Nullable

// App
import jgappsandgames.smartreminderslite.R
import jgappsandgames.smartreminderslite.holder.TaskFolderHolder

// Save
import jgappsandgames.smartreminderssave.tasks.Task

/**
 * TaskAdapterInterface
 * Created by joshua on 12/13/2017.
 *
 * Class that Handles The List View Items For Tasks and Folders Throughout the Entire App
 */
open class TaskAdapterInterface(val activity: Activity, private val listener: TaskFolderHolder.OnTaskChangedListener, val tasks: ArrayList<Task>): BaseAdapter() {
    // Secondary Initializer -----------------------------------------------------------------------
    constructor(activity: Activity, listener: TaskFolderHolder.OnTaskChangedListener, tasks: ArrayList<String>, @Nullable unused: String?):
            this(activity, listener, ArrayList()) {
        for (t in tasks) this.tasks.add(Task(t))
    }

    // List Methods --------------------------------------------------------------------------------
    override fun getCount(): Int {
        return tasks.size
    }

    override fun getViewTypeCount(): Int {
        return 3
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    // Item Methods --------------------------------------------------------------------------------
    override fun getItem(position: Int): Task {
        return tasks[position]
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).getID()
    }

    override fun getItemViewType(position: Int): Int {
        return getItem(position).getType()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        if (convertView == null) {
            val temp: View = if (getItemViewType(position) == Task.TYPE_TASK) LayoutInflater.from(activity).inflate(R.layout.list_task, parent, false)
                else LayoutInflater.from(activity).inflate(R.layout.list_folder, parent, false)

            val holder = TaskFolderHolder(getItem(position), temp, activity, listener)
            temp.tag = holder
            holder.setViews()
            return temp
        }

        val holder: TaskFolderHolder = convertView.tag as TaskFolderHolder
        holder.task = getItem(position)
        holder.setViews()
        return convertView
    }
}